mlir/lib/Analysis/Presburger/Simplex.cpp

```diff
@@ -0,0 +1,929 @@
+//===- Simplex.cpp - MLIR Simplex Class -------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Analysis/Presburger/Simplex.h"
+#include "mlir/Analysis/Presburger/Matrix.h"
+#include "mlir/Support/MathExtras.h"
+
+#include <algorithm>
+#include <cstdio>
+
+namespace mlir {
+const int NULL_INDEX = std::numeric_limits<int>::max();
+
+// Construct a Simplex object with `nVar` variables.
+Simplex::Simplex(unsigned nVar)
+    : nRow(0), nCol(2), tableau(0, 2 + nVar), empty(false) {
+  colVar.push_back(NULL_INDEX);
+  colVar.push_back(NULL_INDEX);
+  for (unsigned i = 0; i < nVar; ++i) {
+    var.push_back(
+        Unknown(/*ownsRow=*/false, /*restricted=*/false, /*pos=*/nCol));
+    colVar.push_back(i);
+    nCol++;
+  }
+}
+
+Simplex::Simplex(const FlatAffineConstraints &constraints)
+    : Simplex(constraints.getNumIds()) {
+  for (unsigned i = 0; i < constraints.getNumInequalities(); ++i)
+    addInequality(constraints.getInequality(i));
+  for (unsigned i = 0; i < constraints.getNumEqualities(); ++i)
+    addEquality(constraints.getEquality(i));
+}
+
+const Simplex::Unknown &Simplex::unknownFromIndex(int index) const {
+  assert(index != NULL_INDEX && "NULL_INDEX passed to unknownFromIndex");
+
+  if (index >= 0)
+    return var[index];
+  else
```

Comment:
Just some nitpicking: According to the llvm coding style guide they prefer to omit an else after a return: https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return